rpc/types/src/string.rs

```diff
@@ -0,0 +1,217 @@
+}
+
+impl rpc_description::HasValueHint for RpcHexString {
+    const HINT: VH = VH::HEX_STRING;
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct HexStringSerde(String);
+
+impl From<RpcHexString> for HexStringSerde {
+    fn from(value: RpcHexString) -> Self {
+        Self(hex::encode(value.0))
+    }
+}
+
+impl TryFrom<HexStringSerde> for RpcHexString {
+    type Error = hex::FromHexError;
+
+    fn try_from(value: HexStringSerde) -> Result<Self, Self::Error> {
+        Ok(RpcHexString::from_bytes(hex::decode(value.0)?))
+    }
+}
+
+/// A binary string type suitable for use in RPC input parameters. Accepts a string or a hex
+/// string.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[serde(into = "RpcStringInSerde", from = "RpcStringInSerde")]
+pub struct RpcStringIn(Vec<u8>);
+
+impl RpcStringIn {
+    pub fn from_bytes(data: Vec<u8>) -> Self {
+        Self(data)
+    }
+
+    pub fn from_string(data: String) -> Self {
+        Self(data.into_bytes())
+    }
+
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.0
+    }
+
+    pub fn try_into_string(self) -> Result<String, (Self, std::str::Utf8Error)> {
+        String::from_utf8(self.0).map_err(|e| {
+            let err = e.utf8_error();
+            let this = Self::from_bytes(e.into_bytes());
+            (this, err)
+        })
+    }
+
+    pub fn try_as_str(&self) -> Result<&str, std::str::Utf8Error> {
+        std::str::from_utf8(self.as_ref())
+    }
+}
+
+impl AsRef<[u8]> for RpcStringIn {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl From<Vec<u8>> for RpcStringIn {
+    fn from(value: Vec<u8>) -> Self {
+        Self::from_bytes(value)
+    }
+}
+
+impl From<String> for RpcStringIn {
+    fn from(value: String) -> Self {
+        Self::from_string(value)
+    }
+}
+
+impl From<RpcHexString> for RpcStringIn {
+    fn from(value: RpcHexString) -> Self {
+        Self::from_bytes(value.into_bytes())
+    }
+}
+
+impl HasValueHint for RpcStringIn {
+    const HINT: VH = VH::Choice(&[&VH::STRING, &VH::Object(&[("hex", &VH::HEX_STRING)])]);
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum RpcStringInSerde {
+    Text(String),
+    Hex(RpcHexString),
+    #[serde(untagged)]
+    Bare(String),
+}
+
+impl From<RpcStringIn> for RpcStringInSerde {
+    fn from(value: RpcStringIn) -> Self {
+        Self::Hex(RpcHexString::from_bytes(value.0))
+    }
+}
+
+impl From<RpcStringInSerde> for RpcStringIn {
+    fn from(value: RpcStringInSerde) -> Self {
+        match value {
+            RpcStringInSerde::Text(s) | RpcStringInSerde::Bare(s) => s.into(),
+            RpcStringInSerde::Hex(hex) => hex.into(),
+        }
+    }
+}
+
+/// A binary string type suitable for use in RPC return values.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[serde(into = "RpcStringOutSerde", try_from = "RpcStringOutSerde")]
+pub struct RpcStringOut(Vec<u8>);
```

Comment:
While the naming is not an important issue now, it's bad for something that goes straight to the blockchain to put "RPC" in it... I would call it something like `WrappedUtf8String`.

Reply:
> While the naming is not an important issue now, it's bad for something that goes straight to the blockchain to put "RPC" in it... I would call it something like `WrappedUtf8String`.

I am not sure what you mean by straight to the blockchain. This is specifically for RPC. It does not have `Encode`/`Decode`. If you want to put it into blockchain, you have to convert it to `Vec<u8>` or something similar.

Reply:
Oh, my mistake. Never mind. I confused it with something else.